src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,609 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for interacting with qemu image builder."""
+
+import dataclasses
+import hashlib
+import logging
+import os
+import shutil
+
+# Ignore B404:blacklist since all subprocesses are run with predefined executables.
+import subprocess  # nosec
+import sys
+import urllib.error
+import urllib.request
+from contextlib import redirect_stdout
+from pathlib import Path
+from typing import Literal
+
+from github_runner_image_builder.chroot import ChrootBaseError, ChrootContextManager
+from github_runner_image_builder.config import Arch, BaseImage
+from github_runner_image_builder.utils import retry
+
+logger = logging.getLogger(__name__)
+
+APT_DEPENDENCIES = [
+    "qemu-utils",  # used for qemu utilities tools to build and resize image
+    "libguestfs-tools",  # used to modify VM images.
+    "cloud-utils",  # used for growpart.
+]
+
+
+# nosec: B603: All subprocess runs are run with trusted executables.
```

Comment:
Are the following `# nosec: B603` comments in this file still needed?

Reply:
I've added this explanation as a header to avoid repeating the messages - but the `# nosec` is applicable only per line unless I disable it in the file entirely (in pyproject.toml or tox.ini) but I thought that would bring risks in the future. 
Would you prefer it be ignored entirely?